### ISSUE TITLE
Modifications to NavBar for Keyboard navigation

### ISSUE THREAD

**Issue by ariel-phet** [phetsims/joist#374]
The navbar is going to require some updates to accommodate a "help button" for keyboard navigation (a keyboard navigable button that will bring up a dialog that shows the various keyboard options available in a keyboard accessible simulation).

The proposed design is:
![navbar_mods](https://cloud.githubusercontent.com/assets/4051174/20574772/779b4a4a-b173-11e6-80a4-d24dcd94054a.png)
 
This design has been made with the possibility of 5 total screens in mind, the proposed changes are:

1. Change the "hamburger" menu button to 3 vertical dots (the PhET logo and 3 dots will all be one button that can open the current menu). This design saves horizontal space and the "3 dot" is currently used in chrome, and seen on mobile browsers
2. Include the option for a "keyboard" icon that will appear if the sim includes keyboard accessibility.
3. Move the homescreen button to the left of the screen selection buttons. This order seems fairly natural since you begin the sim on the home screen. In addition, when 5 screens are present the homescreen button and keyboard help button would be in close proximity, and this arrangement helps not confuse these buttons
4. The homescreen icon has been reduced in size slightly
5. The maximum width of the title may need to be reduced to accommodate these changes in a sim with many screens (4-5)

It will be good to discuss:

- [ ] How invasive will these changes be?
- [ ] Anything in this design we see as a road block that has not yet been anticipated?
- [ ] If we want to keep look and feel consistent, is a maintenence release possible, or will that be a nightmare (we could live with rolling releases and some inconsistency in the navbar, it will be minor for most sims, mainly the new menu icon and the switch in position of the homescreen icon)

**Comment by jonathanolson** [comment]
The changes would probably be fairly isolated in Joist, so it wouldn't be too prohibitive to do maintenance releases.

That said, we may end up with a good number of differing implementations (due to various refactorings that have touched NavigationBar.js and the buttons). I at least did moderate refactoring in August 2015 (https://github.com/phetsims/joist/commit/772985365471edbe7e8c147d577c55702f541df5), and there are a good number of other changes. It will probably require testing the navbar fairly well for 10+ versions (which should be much faster than testing full sims!)

**Comment by samreid** [comment]
The proposed changes sound straightforward.

**Comment by pixelzoom** [comment]
(1) Make sure we consider `brand=adapted-from-phet`, since that logo is much wider than 'brand=phet'.
(2) Sim title in the navbar will likely need to shrink. "Reactants, Products and Leftovers" is probably the current worst-case.

**Comment by samreid** [comment]
We may want to disable the keyboard icon for iPads, it depends on whether.

**Comment by jessegreenberg** [comment]
We decided that @jessegreenberg will work on the implementation, with JO possibly helping where necessary and doing the review.

@arouinfar would you please take a look at the comments in https://github.com/phetsims/joist/issues/374#issuecomment-264261584 to consider the "adapted from PhET" brand case?

**Comment by ariel-phet** [comment]
@arouinfar we need the PhET-iO log to work "as is" -- @jessegreenberg has that logo asset.

For the "adapted from PhET brand" if that is not looking so good (since it is so wide) we will plan to adjust that logo (not the navbar implementation). Perhaps you and I can work on a simple treatment, and then we will give to Dave for refinement.

@jessegreenberg can you let both @arouinfar and I know how many pixels the PhET-iO logo takes up on the navbar so that can be used as the maximum width for any other logos?

**Comment by jonathanolson** [comment]
The phet-io logo (from brand) is 388x108 pixels in size. Logos look like they are sized by their aspect ratio, so if the height is 108 and width is 388 or less, it should be good.

**Comment by arouinfar** [comment]
@ariel-phet I can certainly work on mocking up variations using the other PhET logos.

I'm not finding the PhET-iO or "adapted from PhET" logos in our Design Resources - Branding folder. @jessegreenberg do you have the .ai file for these logos? If you email them to me I can store them on Drive.

**Comment by samreid** [comment]
@arouinfar please look at https://github.com/phetsims/brand/blob/8c6f9c0402d8d445e61b8692c25b78b0ef09f279/phet-io/assets/phet-io-logo.ai

**Comment by samreid** [comment]
And https://github.com/phetsims/brand/tree/dd452990dd83173fb5e6fe29346de9972b428a86/adapted-from-phet/images (but I suspect "adapted from" text may be added in joist code)

**Comment by pixelzoom** [comment]
> (but I suspect "adapted from" text may be added in joist code)

That is indeed the case, since it's translatable.  See joist.AdaptedFromText.

**Comment by arouinfar** [comment]
Thanks @samreid and @pixelzoom!

**Comment by jessegreenberg** [comment]
This is mostly complete, commit coming shortly.  While aqua is running tests here are some screenshots with these changes.  I think the keyboard navigation icon was still under consideration in https://github.com/phetsims/joist/issues/374#issue-191355797 so I used one from FontAwesome for now.

Proposed mockup against navigation bar in sim (mockup on top):
![capture2](https://cloud.githubusercontent.com/assets/6396244/20955735/bcf9bdd0-bc02-11e6-8356-3b87aac889b2.PNG)

Here is a "worst case" scenario with a pretty long name, 5 screens, keyboard help button, and "Adapted From" logo:
![capture3](https://cloud.githubusercontent.com/assets/6396244/20955927/ffcfc54a-bc03-11e6-8f94-303bf3694d13.PNG)

Here are a couple examples without the Keyboard Help Button, which is how the navigation bar will look for most sims for quite a while:
![capture4](https://cloud.githubusercontent.com/assets/6396244/20955886/bbbdfd72-bc03-11e6-9fdf-225b664b842a.PNG)
![capture7](https://cloud.githubusercontent.com/assets/6396244/20955893/c7bc8724-bc03-11e6-9e60-50d8dab684fb.PNG)

**Comment by samreid** [comment]
Looking great so far @jessegreenberg!  Will you please align the bottom of the house to the bottom of the screen icon rectangles?  I think it will look a bit nicer.

**Comment by pixelzoom** [comment]
Have you tested with `stringTest=long` ?

**Comment by pixelzoom** [comment]
@samreid asked:
> Will you please align the bottom of the house to the bottom of the screen icon rectangles? 

I'm guessing that might be non-trivial to do. The screen labels are scaled to occupy a maximum width, which also affects the height of the labels.  And the icon+label is vertically centered in the navbar.  So the position of the bottom of the screen icons is highly dependent on the labels.

**Comment by pixelzoom** [comment]
And imo, the home screen icon looks better vertically centered anyway.

**Comment by arouinfar** [comment]
Thanks @jessegreenberg! The worst-case scenario isn't as bad as I expected. I think we can make a few tweaks and avoid a complete overhaul of the "adapted from" logo. I'll work on a mockup, but here's what I have in mind:
* The "adapted from" text could be made 1 or 2 pt smaller. It feels pretty big compared to the other text in the navbar.
* The title looks a bit too small in the worst case scenario. Centering the home screen icon + screen icons in the navbar (instead of just the screen icons) would give the title a bit more room. 
* If we center the home screen icon + screen icons in the navbar, we'll lose some negative space to the right of the last screen icon. Depending on how this looks, we could consider slightly reducing the padding between the keyboard icon and the logo.

> And imo, the home screen icon looks better vertically centered anyway.

I agree with @pixelzoom here. I think if we bottom-align the home screen icon with the screen icons, it will bee too close to the top of the navbar. My original mockups had a bit more black space on the top of the screen icons, so a bottom-aligned home screen icon looked more reasonable than I think it would look in reality. Given that bottom-alignment is non-trivial, I would suggest we leave it centered.

**Comment by samreid** [comment]
I see the point about the top of the house being too close to the top of the navbar, thanks @arouinfar!

**Comment by arouinfar** [comment]
<img width="1041" alt="screen shot 2016-12-07 at 11 38 28 am" src="https://cloud.githubusercontent.com/assets/8419308/20981377/b5b72dc2-bc71-11e6-94ff-9e493ece41ce.png">



Top: @jessegreenberg's worst-case scenario from https://github.com/phetsims/joist/issues/374#issuecomment-265360377
Middle: worst-case scenario with tweaks suggested in https://github.com/phetsims/joist/issues/374#issuecomment-265520312. Specifically:
* Grouped home screen icon with the screen icons and centered it within the navbar
* Reduced "adapted from" text to 12 pt, and moved 2 px closer to PhET logo
* Spacing between keyboard icon and edge of logo is 25 px

Bottom: Same as middle, but a more typical scenario with a shorter sim name and 3 screens.

Here's what it would look like if the screen names are longer stings
<img width="1035" alt="screen shot 2016-12-07 at 11 58 11 am" src="https://cloud.githubusercontent.com/assets/8419308/20982138/871ba9b8-bc74-11e6-8d98-a74609e6c233.png">


I don't think the my last mockup is particularly realistic, but I think it's something we could live with. In general, I think the tweaks I've made to @jessegreenberg's worst-case scenario are an improvement.  What do you think @ariel-phet?

**Comment by pixelzoom** [comment]
And to ask again... this is all being tested with "stringTest=long", yes?

**Comment by pixelzoom** [comment]
And to ask again... this is all being tested with `stringTest=double` and `stringTest=long`, yes?

**Comment by arouinfar** [comment]
If I had access to a string-testable version, I would @pixelzoom. The 2nd mockup in https://github.com/phetsims/joist/issues/374#issuecomment-265543331 is about the best I can do in Illustrator.

**Comment by arouinfar** [comment]
As it currently stands, I have no idea if/where I can access @jessegreenberg's work on phettest.

**Comment by jessegreenberg** [comment]
> This is all being tested with stringTest=double and stringTest=long, yes?

Yes, here are stringTest=long and stringTest=double tests with the layout before the additional changes in https://github.com/phetsims/joist/issues/374#issuecomment-265543331.

![capture9](https://cloud.githubusercontent.com/assets/6396244/20984570/1e304bf2-bc7e-11e6-8d4d-9a0072be5718.PNG)
![capture10](https://cloud.githubusercontent.com/assets/6396244/20984572/1f53f72c-bc7e-11e6-9630-0ba1777da010.PNG)

@arouinfar I will push to master for phettest once the changes in https://github.com/phetsims/joist/issues/374#issuecomment-265543331 are done.

**Comment by pixelzoom** [comment]
Looks great. I asked because the last time I was modifying the navabar, `stringTest` was critical to verifying robustness of the layout.

**Comment by jessegreenberg** [comment]
Absolutely, ```stringTest``` has caught a few issues with my layout for changes for this issue!

The above commits are the work for this issue thus far, including the list in https://github.com/phetsims/joist/issues/374#issuecomment-265543331.  Here are some examples:

Comparing mockups from @arouinfar (mock ups on top):
![capture1](https://cloud.githubusercontent.com/assets/6396244/20989209/231ae61e-bc91-11e6-8266-1a0ee98e038e.PNG)
![capture4](https://cloud.githubusercontent.com/assets/6396244/20989213/2531083e-bc91-11e6-8e56-f0d6794c9b20.PNG)

String tests:
![capture2](https://cloud.githubusercontent.com/assets/6396244/20989222/2cc7952c-bc91-11e6-8cc3-a42ac5419a8a.PNG)
![capture3](https://cloud.githubusercontent.com/assets/6396244/20989232/35c96196-bc91-11e6-8397-520c81fec498.PNG)

Single Screen:
![capture6](https://cloud.githubusercontent.com/assets/6396244/20989242/3dc08a78-bc91-11e6-9ac6-6aec2237d689.PNG)

**Comment by jessegreenberg** [comment]
@arouinfar @ariel-phet @emily-phet can you please review these?  The button will be hidden (since most sims aren't accessible and there is no content yet for the "Keyboard Help Dialog").  But you should be able to find changes on phettest to review the new layout.

We still need to come up with an icon for this button.

**Comment by zepumph** [comment]
Sorry about that mumbo jumbo, I accidentally committing some incorrect files. All is well now

**Comment by emily-phet** [comment]
@jessegreenberg where am I supposed to be looking? I'm a bit confused.

**Comment by jessegreenberg** [comment]
@emily-phet, the images in https://github.com/phetsims/joist/issues/374#issuecomment-265593015 show what the navigation bar looks like now with the keyboard help button for sims with 5, 3, and 1 screens.

With phettest, @arouinfar can pull the latest version of joist and see the changes directly in a sim, but I am not sure how to set that up.

**Comment by arouinfar** [comment]
@jessegreenberg the screenshots in https://github.com/phetsims/joist/issues/374#issuecomment-265593015 look good to me. The slight differences between the mockups and your screenshots are errors on my end in trying to approximate the navbar layout. 

Unfortunately, I can't take a look at the navbar in context, because common libraries cannot be viewed directly on phettest. Perhaps I can take a look on your machine tomorrow before/after status?

**Comment by samreid** [comment]
@arouinfar I encourage you to look at this with @jessegreenberg, but also wanted to let you know that after pressing "pull all" on phettest, you can launch any sim to see the changes in that sim.

**Comment by arouinfar** [comment]
Thanks for the clarification @samreid! After pulling all, I took a look at FAMB with `?stringTest=double` and the navbar still looks reasonable to me. 
<img width="1072" alt="screen shot 2016-12-07 at 6 09 21 pm" src="https://cloud.githubusercontent.com/assets/8419308/20993564/6303ca5e-bca8-11e6-8b75-0e0a9ab6d0d1.png">

I also tried `?stringTest=Beschleunigung` to see how the screen icons would look with a long German name.
<img width="1071" alt="screen shot 2016-12-07 at 6 10 48 pm" src="https://cloud.githubusercontent.com/assets/8419308/20993601/982be176-bca8-11e6-81f1-e0e068ae3d05.png">

I'd still like to see a navbar with a keyboard icon, so perhaps @jessegreenberg can show me one tomorrow.

**Comment by ariel-phet** [comment]
@arouinfar my suggestion for a revised "adapted from PhET"

**Comment by ariel-phet** [comment]
@arouinfar A thought I was having that might simplify the "adapted from" logo a bit. I took the "PhET-iO" logo chopped off the "iO" -- this has the advantage of make our name clear (for those less familiar) and reduces the vertical height of the logo. I feel like without the swoop and the airplane it looks less like it comes directly from us. I also tried the words "adapted from" on top. I think in this branding, it is OK for out logo to be a bit smaller.

![adapted](https://cloud.githubusercontent.com/assets/4051174/20995763/60760f28-bcb6-11e6-869e-58c0f8829732.PNG)

**Comment by emily-phet** [comment]
@jessegreenberg wrt the the keyboard nav button, the mockups all look good to me.

**Comment by arouinfar** [comment]
@ariel-phet's suggested changes to the logo clean up the navbar quite nicely. Let's take a look after status, and I can send @jessegreenberg the logo asset if it gets the thumbs up.

<img width="1031" alt="screen shot 2016-12-08 at 10 12 52 am" src="https://cloud.githubusercontent.com/assets/8419308/21019767/f09b8e66-bd2e-11e6-8482-136ff133ce27.png">

**Comment by jessegreenberg** [comment]
That looks good to me, and it saves us a bit of space.

**Comment by arouinfar** [comment]
I showed the updated adapted-from-PhET logo to @kathy-phet and she's given it her approval. 

@ariel-phet do you have any further tweaks to suggest?

@jessegreenberg how would you prefer the asset? Should I send you the .ai file for just the PhET portion of the logo?

**Comment by ariel-phet** [comment]
@arouinfar you should ask @kathy-phet if that version should have the registered "R" or not.

I would also do the programming such that the "adapted from" max width is just the width of the logo

**Comment by kathy-phet** [comment]
yes at the "R".

**Comment by arouinfar** [comment]
Thanks @kathy-phet @ariel-phet. I'll add the R to the asset I send to @jessegreenberg.

**Comment by arouinfar** [comment]
@ariel-phet here are three versions of the registered adapted-from-PhET logo. The R symbol is the same size as it normally is. The PhET portion of the logo is a bit smaller in this iteration, so the R is a bit too tight if it is bottom- and right-aligned with the T, like it usually is (top). I think it looks a bit cleaner if the R is moved one pixel further to the right (middle). Alternatively, we could center-align the R symbol with the end of the T (bottom). I'm not a fan of the last option, but could live with it. 

Which do you prefer @ariel-phet? 

![screen shot 2016-12-12 at 1 09 47 pm](https://cloud.githubusercontent.com/assets/8419308/21114969/5875b10c-c06c-11e6-84e6-fb881ac2c14c.png)

**Comment by arouinfar** [comment]
Spoke with @ariel-phet, and we've decided to go with the middle option. I've emailed the assets to @jessegreenberg for implementation.

**Comment by jessegreenberg** [comment]
I updated the adapted from phet logo in the above commits.  Here are a couple screenshots:

![normal](https://cloud.githubusercontent.com/assets/6396244/21159202/8aa6f7b8-c13d-11e6-9dd9-bb0f4e04df34.PNG)

![string-test-long](https://cloud.githubusercontent.com/assets/6396244/21159207/8e15a52a-c13d-11e6-9d14-9e19d626529a.PNG)

@arouinfar can you please review?

**Comment by arouinfar** [comment]
@jessegreenberg your screenshots look good to me. I'd review on phettest, but it appears to be down.

**Comment by arouinfar** [comment]
>  I'd review on phettest, but it appears to be down.

Scratch that, apparently Chrome was due for a hard refresh... 

I'm taking a look now.

**Comment by arouinfar** [comment]
@jessegreenberg still looks good on master!

At one point we had talked about grouping the sim name and the home button into a single click/touch area that would navigate to the home screen. @ariel-phet is this something we still want to do?

**Comment by ariel-phet** [comment]
@arouinfar no -- I think the devs generally felt best to just keep the home screen icon as its own separate element for now.

**Comment by arouinfar** [comment]
Sounds good @ariel-phet. Thanks!

**Comment by jessegreenberg** [comment]
Alright, thanks!  I think we are finished with changes for this issue then.  @jonathanolson would you please review changes?  john-travoltage will be the first sim to use this, so there might be some overlap between this issue and phetsims/john-travoltage#128.

**Comment by samreid** [comment]
@jessegreenberg and @ariel-phet: in https://github.com/phetsims/joist/issues/383 we decided to make the "adapted from PhET" logo a single non-translatable image and show it in the splash screen (as well as the navbar).  I was going to proceed with this, but it looks like the AI files don't include the "adapted from" text, perhaps it is still added in Joist.  @ariel-phet and @jessegreenberg how would you like to proceed?

**Comment by arouinfar** [comment]
@samreid I have the .ai asset with the "adapted from" text. I will email it to you.

**Comment by jessegreenberg** [comment]
Thanks @arouinfar.  @samreid I could also update the image and assets for 'adapted-from-phet' if you wish.

**Comment by samreid** [comment]
I'll be starting on this now, and I'll report back if I need assistance, thanks!

**Comment by samreid** [comment]
Completed above, @jessegreenberg can you please review?

**Comment by jessegreenberg** [comment]
Looks great to me, thanks @samreid.  In the above commit I added the dialog that will open from this button.

Assigning back to @jonathanolson for review.

**Comment by jonathanolson** [comment]
Usage of the pattern (2 usages):
```js
sim.lookAndFeel.navigationBarFillProperty.linkAttribute( this.background, 'fill' );
```
can be inlined in the options:
```js
{
  // ...
  fill: sim.lookAndFeel.navigationBarFillProperty
}
```


If KeyboardHelpButton is already being passed the full {Sim}, can it pull out `sim.lookAndFeel.navigationBarFillProperty` on its own?


In KeyboardHelpButton, it's doing a multilink to that handles light/dark:
```js
Property.multilink( [ backgroundFillProperty, sim.showHomeScreenProperty ],
  function( backgroundFill, showHomeScreen ) {
    var backgroundIsWhite = backgroundFill !== 'black' && !showHomeScreen;
    icon.image = backgroundIsWhite ? darkIconMipmap : brightIconMipmap;
  } );
```
This seems duplicated with LookAndFeel code:
```js
this.navigationBarFillProperty = new DerivedProperty( [ this.backgroundColorProperty ],
  function( backgroundColor ) {
    var screenIsBlack = !!new Color( backgroundColor ).equals( Color.BLACK );
    return screenIsBlack ? 'white' : 'black';
  } );
```
so presumably we should have a BooleanProperty to determine whether the navbar background is dark/light, and have multiple properties derived from it (like the navbar fill AND the keyboard icon image). Also note the Color.equals() comparison, as '#000', '#000000', 'rgb(0,0,0)', 'rgba(0,0,0,1)', 'hsl(X,X,0)' and 'hsla(X,X,0,1)' are all black, but not === to 'black'.


I saw `!platform.mobileSafari` as a check for the keyboard node. Do we also want to check mobile in general (e.g. Android)?


For determining available layout width, e.g.:
```js
// available width right of center
var availableRight = ( HomeScreenView.LAYOUT_BOUNDS.width / 2 ) - this.phetButton.width - PHET_BUTTON_RIGHT_MARGIN -
                       this.keyboardHelpButton.width - KEYBOARD_HELP_BUTTON_LEFT_MARGIN;
```
I don't see PHET_BUTTON_LEFT_MARGIN used, but for positioning the keyboard icon I see:
```js
this.keyboardHelpButton.right = this.phetButton.left - PHET_BUTTON_LEFT_MARGIN;
```
Maybe horizontal documentation that is graphical would help (or is this a bug?):

Horizontally:
```
TITLE_LEFT_MARGIN
{title}
TITLE_RIGHT_MARGIN
{
  PHET_BUTTON_LEFT_MARGIN
  {PhetButton}
  PHET_BUTTON_RIGHT_MARGIN
  [ SCREEN_BUTTON_SPACING ] when in between
}*
[
  KEYBOARD_HELP_BUTTON_LEFT_MARGIN
  {keyboard help button}
]
HOME_BUTTON_LEFT_MARGIN
{Home button}
HOME_BUTTON_RIGHT_MARGIN
```
is that correct?

**Comment by jonathanolson** [comment]
Also leaving myself assigned, as there's further review once confirming that is how the margin is ordered.

**Comment by jessegreenberg** [comment]
Yikes, sorry this fell of the radar, ill be working on this soon.

**Comment by zepumph** [comment]
This works out, because I would love the review to happen after my commit from today for https://github.com/phetsims/joist/issues/454 (specifically https://github.com/phetsims/joist/commit/b40fa583ef1d181daff111cecd879e03adcfa4cc) that refactors the keyboard help button's layout if it isn't added.

**Comment by pixelzoom** [comment]
Raising priority to top.  Apparently there were changes to the navbar made in March _2017_ that have not been reviewed.

**Comment by pixelzoom** [comment]
@jessegreenberg @zepumph who is supposed to be reviewing this?

**Comment by jessegreenberg** [comment]
I am meant to work on this.

**Comment by jessegreenberg** [comment]
I made the above commits to address comments in https://github.com/phetsims/joist/issues/374#issuecomment-277079419. Responding to  each comment:

> Usage of the pattern (2 usages)...can be inlined in the options:

Done.

>If KeyboardHelpButton is already being passed the full {Sim}, can it pull out sim.lookAndFeel.navigationBarFillProperty on its own?

Done.

> presumably we should have a BooleanProperty to determine whether the navbar background is dark/light, and have multiple properties derived from it (like the navbar fill AND the keyboard icon image). 

Done.

> I saw !platform.mobileSafari as a check for the keyboard node. Do we also want to check mobile in general (e.g. Android)?

No, there was a bug with Apple bluetooth keyboards where certain keyboard event properties were not passed to the browser, making it impossible to use some sim features. We verified this was not a problem for Android and submitted a bug report to Apple.

> Maybe horizontal documentation that is graphical would help

I added something to this effect around the constants in NavigationBar.js.

@jonathanolson (and others) I am sorry this was forgotten and took so long. Could you please review the above commits in response to https://github.com/phetsims/joist/issues/374#issuecomment-277079419?

**Comment by pixelzoom** [comment]
This is labeled for "meeting:developer" because it involves significant changes to navigation bar, was assigned for review on 12/16/2017, and has been labeled "top priority" for 16 days.

**Comment by jonathanolson** [comment]
@jessegreenberg's commits above look good, sorry about the delay.

**Comment by jessegreenberg** [comment]
Review was the final  portion of this issue. This issue can be closed.

At 4/12/18 developer meeting we talked about how to handle and prioritize review issues. @ariel-phet said that for review issues that will take 10-30 minutes, go ahead and assign to a developer . @ariel-phet said that for larger issues we should assign to him to assign a developer and prioritize. But we should generally use our judgement for whether or not it needs a thorough review.

@zepumph mentioned that in the review request, it is very helpful to add notes about what specifically was changed, and what in particular is questionable.
